python/lsst/ts/ledprojector/ledprojector_csc.py

```diff
@@ -0,0 +1,336 @@
+    version = __version__
+    valid_simulation_modes = (0, 1)
+
+    def __init__(
+        self,
+        config_dir=None,
+        initial_state=salobj.State.STANDBY,
+        override="",
+        simulation_mode=0,
+    ):
+        self.led_model = None
+
+        # Unit tests can set either or both of these true to make
+        # the connect method time out in the appropriate model.
+        # Ignored unless in simulation mode.
+        self.led_make_connect_time_out = False
+
+        # Set True just after the led are connected,
+        # and false just before disconnecting.
+        self.should_be_connected = False
+
+        self.config = None
+
+        super().__init__(
+            name="LEDProjector",
+            index=0,
+            config_schema=CONFIG_SCHEMA,
+            config_dir=config_dir,
+            initial_state=initial_state,
+            override=override,
+            simulation_mode=simulation_mode,
+        )
+
+    @staticmethod
+    def get_config_pkg():
+        return "ts_config_atcalsys"
+
+    @property #add index
+    def led_connected(self):
+        """Return True if the LabJack is connected and has seen status."""
+        return self.led_model is not None and self.led_model.status_seen
+
+#configure should include either array of configs, or config contains array for LEDs
+    async def configure(self, config):
+        config = copy.copy(config)
+        config.led = types.SimpleNamespace(**config.led)
+        self.config = config
+
+#from our perspective, there can only be the 'led model' as one object, since the only failure point is the labjack
+#the labjack then connects, effectively one way, to the LEDs, so from this objects perspective we only have on, off, and no failure states
+#the failure state lies in the labjack not connecting/responding
+    async def handle_summary_state(self):
+        if self.disabled_or_enabled:
+            if not self.led_connected:
+                try:
+                    await self.connect_led()
+                except Exception as e:
+                    return await self.fault(
+                        code=ErrorCode.led_ERROR,
+                        report=f"Could not connect to the led: {e!r}",
+                    )
+            self.should_be_connected = True
+        elif self.summary_state == salobj.State.FAULT:
+            if self.led_model and self.led_model.led_was_on:
+                # Turn off the led, if connected
+                if self.led_connected:
+                    self.log.warning(
+                        "Going to fault while connected to the led controller; "
+                        "trying to turn the led off immediately."
+                    )
+                    try:
+                        await self.led_model.turn_led_off(
+                            force=True, wait=False, reason="CSC is going to FAULT state"
+                        )
+                    except Exception as e:
+                        self.log.error(
```

Comment:
Use `self.log.exception` to get the full traceback. Then you can remove the `{e!r}` part of the error message and drop the `as e` above, e.g.:

```
except Exception:
    self.log.exception("Failed to turn LED off. You may have to turn it off manually.")
```